pretrain/openwebtext_trainer.py

```diff
@@ -0,0 +1,210 @@
+        )
+
+    def on_train_batch_start(self, batch: Any, batch_idx: int) -> None:
+        if not decay_lr:
+            return
+        # determine and set the learning rate for this iteration
+        lr = get_lr(self.trainer.fit_loop.total_batch_idx)
+        for optimizer in self.trainer.strategy.optimizers:
+            for param_group in optimizer.param_groups:
+                param_group["lr"] = lr
+
+    def training_step(self, batch: Any, batch_idx: int) -> torch.Tensor:
+        input_ids, targets = batch
+        logits = self.module(input_ids)
+        loss = chunked_cross_entropy(logits, targets, chunk_size=0)
+        self.log("train_loss", loss, on_step=True, on_epoch=False, prog_bar=True)
+        return loss
+
+    def validation_step(self, batch: Any, batch_idx: int) -> torch.Tensor:
+        input_ids, targets = batch
+        logits = self.module(input_ids)
+        loss = chunked_cross_entropy(logits, targets, chunk_size=0)
+        self.log("val_loss", loss, on_step=False, on_epoch=True, prog_bar=True)
+        return loss
+
+
+def main(devices: int = 1, precision: Optional[str] = None, tpu: bool = False) -> None:
+    if precision is None:
+        precision = "32-true" if tpu else "bf16-mixed"
+    if devices > 1:
+        if tpu:
+            # For multi-host TPU training, the device count for Fabric is limited to the count on a single host.
+            devices = "auto"
+            strategy = XLAStrategy(sync_module_states=False)
+        else:
+            auto_wrap_policy = partial(transformer_auto_wrap_policy, transformer_layer_cls={Block})
+            strategy = FSDPStrategy(
+                auto_wrap_policy=auto_wrap_policy,
+                activation_checkpointing=Block,
+                # the argument is not available in the Trainer strategy, but it's the default anyways
+                # state_dict_type="full",
+                limit_all_gathers=True,
+                cpu_offload=False,
+            )
+    else:
+        strategy = "auto"
+
+    logger = step_csv_logger("out", name, cls=CSVLogger, flush_logs_every_n_steps=log_interval)
+    speed_monitor = SpeedMonitorCallback(
+        length_fn=lambda batch: batch[0].size(1), batch_size=micro_batch_size, window_size=50, time_unit="seconds"
+    )
+    model_checkpoint = ModelCheckpoint(dirpath=out_dir, every_n_train_steps=save_interval, verbose=True)
+    trainer = Trainer(
+        devices=devices,
+        strategy=strategy,
+        precision=precision,
+        logger=logger,
+        callbacks=[speed_monitor, model_checkpoint],
+        max_steps=max_iters,
+        max_epochs=1,
+        limit_val_batches=eval_iters,
+        accumulate_grad_batches=gradient_accumulation_steps,
+        log_every_n_steps=log_interval,
+        val_check_interval=eval_interval,
+    )
+
+    L.seed_everything(1337 + trainer.global_rank, workers=True)
+
+    trainer.print(hparams)
```

Comment:
Why not just do if `global_rank == 0: print`? You can always update the script later. This would remove the need to subclass the Trainer just for this .

Reply:
I'm adding `trainer.print` in https://github.com/Lightning-AI/lightning/pull/17980